File: csharp/P0001_TwoSum/Solutions/HashMapTwoPass.cs
// Time complexity: O(n)
// Space complexity: O(n)

namespace P0001_TwoSum.Solutions.HashMapTwoPass;

public class Solution
{
    public int[] TwoSum(int[] nums, int target)
    {
        Dictionary<int, int> seen = new Dictionary<int, int>();

        // Populate hashmap
        for (int i = 0; i < nums.Length; i++)
            seen[nums[i]] = i;

        // Locate complement in hashmap
        for (int i = 0; i < nums.Length; i++)
        {
            int complement = target - nums[i];

            if (seen.ContainsKey(complement) && seen[complement] != i)
                return [i, seen[complement]];
        }

            // This should be unreachable
            return [-1];
    }
}

File: rust/p0001_two_sum/src/hashmap_two_pass.rs
// Time complexity: O(n)
// Space complexity: O(n)

use std::collections::HashMap;

pub struct Solution;

impl Solution {
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
        let mut seen: HashMap<i32, usize> = HashMap::new();

        for (i, &num) in nums.iter().enumerate() {
            seen.insert(num, i);
        }

        for (i, &num) in nums.iter().enumerate() {
            let complement = target - num;

            if let Some(&j) = seen.get(&complement) {
                if j != i {
                    return vec![i as i32, j as i32];
                }
            }
        }

        unreachable!("This should not be reachable with valid input");
    }
}
